sigstore/verify/verifier.py

```diff
@@ -108,6 +118,76 @@ def _from_trust_config(cls, trust_config: ClientTrustConfig) -> Verifier:
             trusted_root=trust_config.trusted_root,
         )
 
+    def _verify_signed_timestamp(
+        self, timestamp_response: TimeStampResponse, signature: bytes
+    ) -> bool:
+        """
+        Verify a Signed Timestamp using the TSA provided by the Trusted Root.
+        """
+        cert_authorities = self._trusted_root.get_timestamp_authorities()
+        for certificate_authority in cert_authorities:
+            certificates = certificate_authority.certificates(allow_expired=True)
+
+            builder = VerifierBuilder()
+            for certificate in certificates[:-1]:
+                builder.add_intermediate_certificate(certificate)
+            builder.add_root_certificate(certificates[-1])
+
+            verifier = builder.build()
+            try:
+                verifier.verify(timestamp_response, signature)
+            except Rfc3161VerificationError as e:
+                _logger.debug("Unable to verify Timestamp with CA.")
+                _logger.exception(e)
+                continue
+
+            if (
+                certificate_authority.validity_period_start
+                and certificate_authority.validity_period_end
+            ):
+                if (
+                    certificate_authority.validity_period_start
+                    <= timestamp_response.tst_info.gen_time
+                    < certificate_authority.validity_period_end
+                ):
+                    return True
+
+                _logger.debug(
+                    "Unable to verify Timestamp because not in CA time range."
+                )
+            else:
+                _logger.debug(
+                    "Unable to verify Timestamp because no validity provided."
+                )
+
+        return False
+
+    def _verify_timestamp_authority(self, bundle: Bundle) -> int:
```

Comment:
I think this needs to return the verified timestamp, not the _number_ of verified timestamps.

(It needs to return the verified timestamp, because that becomes the source of trusted time for the rest of the verification flow.)

Reply:
Great catch - I've reworked a bit this part to have a more informative return type and integrated the certificate chain validity check.